### PR TITLE
storage_volume: Fix bugs in importing ISO volume

### PR DESCRIPTION
**Pull request by wfdewith**
Trying to import an ISO file through the `source_file` attribute in `resource "incus_storage_volume"` causes the following errors.
```
resource_storage_volume_test.go:317: Step 1/2 error: Error running apply: exit status 1

    Error: Provider produced inconsistent result after apply

    When applying changes to incus_storage_volume.volume1, provider
    "provider[\"registry.terraform.io/hashicorp/incus\"]" produced an unexpected
    new value: .config: new element "size" has appeared.

    This is a bug in the provider, which should be reported in the provider's own
    issue tracker.

    Error: Provider produced inconsistent result after apply

    When applying changes to incus_storage_volume.volume1, provider
    "provider[\"registry.terraform.io/hashicorp/incus\"]" produced an unexpected
    new value: .content_type: was cty.StringVal("filesystem"), but now
    cty.StringVal("iso").

    This is a bug in the provider, which should be reported in the provider's own
    issue tracker.
```

The first error is caused by `config.size` not being considered a computed value. I'm not 100% certain my solution is correct, because I don't fully understand why this is only a problem for ISO files. Why do block volumes not have a size associated with them by default?

As for the second error, the `incus_storage_volume` resource only supports content types of either filesystem or block.

When importing a volume from a file or a source volume, the user should not be allowed to choose the content type, since the content type is entirely dependent on the source file or volume. Therefore, the `content_type` attribute should be exclusive with both `source_file` and `source_volume`. This also means that we can no longer use a default value, since when `source_file` or `source_volume` are used, the `content_type` is a pure computed value and cannot have a default.

Unfortunately, it seems that Terraform does not have straightforward support for setting default values conditionally, so I implemented this with a plan modifier. I'm not very experienced with writing Terraform providers, so if there is a better solution, I'm open to any suggestions.

### CODE REVIEW

**Comment by wfdewith** [comment]
Sure, here you go:
```hcl
terraform {
  required_providers {
    incus = {
      source  = "lxc/incus"
      version = "0.3.1"
    }
  }
}

provider "incus" {
  remote {
    name    = "local"
    scheme  = "unix"
    address = ""
    default = true
  }
}

resource "incus_storage_volume" "some_iso" {
  name        = "some_iso"
  pool        = "default"
  source_file = "some_iso.iso"
}
```

It doesn't really matter what `some_iso.iso` is, I created it with `head -c 256 /dev/zero > some_iso.iso`.

**Comment by wfdewith** [comment]
The following also reproduces the problem for backup files.

* Create a volume with non-default configuration: `incus storage volume create default test --type block size=1GiB`
* Export the config to backup tarball: `incus storage volume export default test`
* Import the backup file through Terraform:
```hcl
terraform {
  required_providers {
    incus = {
      source  = "lxc/incus"
      version = "0.3.1"
    }
  }
}

provider "incus" {
  remote {
    name    = "local"
    scheme  = "unix"
    address = ""
    default = true
  }
}

resource "incus_storage_volume" "some_backup" {
  name        = "some_disk"
  pool        = "default"
  source_file = "backup.tar.gz"
}
```

```hcl

```

**Comment by maveonair** [comment]
@wfdewith Thanks a lot!

First, you need to set the necessary `content_types` for `iso` and `block` to make it work as `filesystem` is the default value:

```hcl
resource "incus_storage_volume" "some_iso" {
  name         = "some_iso"
  pool         = "default"
  source_file  = "some_iso.iso"
  content_type = "iso"
}

resource "incus_storage_volume" "some_backup" {
  name         = "some_disk"
  pool         = "default"
  source_file  = "backup.tar.gz"
  content_type = "block"
}
```

As you mentioned, the `size` should be a computed key and `iso` should be an allowed `content_type`. So, the actual difference should be just this:

```diff
diff --git i/internal/storage/resource_storage_volume.go w/internal/storage/resource_storage_volume.go
index cae45ce..3b94ac9 100644
--- i/internal/storage/resource_storage_volume.go
+++ w/internal/storage/resource_storage_volume.go
@@ -107,7 +107,7 @@ func (r StorageVolumeResource) Schema(_ context.Context, _ resource.SchemaReques
                                        stringplanmodifier.RequiresReplace(),
                                },
                                Validators: []validator.String{
-                                       stringvalidator.OneOf("filesystem", "block"),
+                                       stringvalidator.OneOf("filesystem", "block", "iso"),
                                },
                        },
 
@@ -552,6 +552,7 @@ func (_ StorageVolumeModel) ComputedKeys() []string {
        return []string{
                "block.filesystem",
                "block.mount_options",
+               "size",
                "volatile.",
        }
 }
```

@stgraber I would suggest that I create a new PR with the small changes required and I update the documentation to have a `block` and `iso` example. What do you think?

**Comment by wfdewith** [comment]
This is what I started out with, but I believe it is not sufficient. If you create and export the test volume with other `config` keys, this breaks.
* `incus storage volume create default test --type block size=1GiB security.shared=true`
* `incus storage volume export default test`
* `terraform apply` as you described with your patch.
This results in the following error.
```
╷
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to incus_storage_volume.some_backup, provider
│ "provider[\"registry.terraform.io/lxc/incus\"]" produced an unexpected new value: .config: new element
│ "security.shared" has appeared.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
```

Now, technically, in this situation, the provider isn't really missing functionality. You can add a `config` block to the resource to make it work again:
```hcl
resource "incus_storage_volume" "some_backup" {
  name         = "some_disk"
  pool         = "default"
  source_file  = "backup.tar.gz"
  content_type = "block"
  config = {
    "security.shared" = true
  }
}
```

---

However, I'm arguing that this is undesirable behavior. If you specify any of the following attributes incorrectly when you specified a `source_file`, you get an error.
* `type` (is probably always `custom`, or are there other types?)
* `content_type`
* `description`
* `config`

The error you get is similar to the one above and mentions `This is a bug in the provider, which should be reported in the provider's own issue tracker.`, which is confusing at best.

I see two possible solutions for this problem:
* Introduce an `incus_imported_storage_volume` resource where the user cannot configure the four attributes mentioned. This resource would only include them as outputs.
* Make `source_file` conflict with the four attributes in `incus_storage_volume`, so that you cannot specify both. This requires some extra tweaks, because the four attributes should have default values _unless_ `source_file` is specified. The reason for that is that Terraform uses defaults unconditionally and will error if this default value differs from what was computed afterwards (i.e. in our case what the Incus server returns). From my searching around, the accepted way to create conditional defaults is with plan modifiers, like I did in this PR.

My preference would be the second option, even though it might lead to slightly more complex code, it reduces duplication and keeps the user interface cleaner.

---

I haven't tested this, but everything I've mentioned in this comment is also true if you substitute `source_file` with `source_volume`, based on reading the source code of the provider.

This problem might also exist for other resources that can be imported from a file or copied from an existing object, but I haven't looked at those (yet).

**Comment by maveonair** [comment]
Thanks a lot for your detailed description, and you are right. This requires a more complex solution.

**Comment by wfdewith** [comment]
I was in the process of figuring it out myself so my earlier comments were quite unclear, so I can understand the confusion.

I think I have a good idea for a nice solution and I'll also extend the acceptance tests a bit more to make sure we cover the edge cases as well. I just need to check the other resources to see if this problem exists there as well. I'll open a draft PR for this.
